Add Case Study evaluation ADR (#1)
joaomo/Certify-Architecture-Kata: ADRs/002_adr_architecture_submission.md

# ADR: AI-Assisted Grading of Architecture Submissions

## ADR ID: 002

## Date: 2025-02-14

## Status: Proposed

---

## 1. Introduction

> This ADR proposes implementing and AI-Assisted grading system for architecture test submissions. It would use a hybrid approach with the "human in the loop". It would also include the feedback generation using a similar approach.

Reviewer: ```suggestion
This ADR proposes implementing an AI-Assisted grading system for architecture test submissions. It would use a hybrid approach with the "human in the loop". It would also include the feedback generation using a similar approach.
```

        K[(Case Study DB <br> *see Admin*)] --> E
        B --> |Candidate| F[Candidate Status Service <br> *Candidate Progress* <br> *see Test 1*]
        B --> |Candidate| G[Architecture Submission Service <br> *Architecture Submission*] --> F
        B --> |Expert| H[Architecture Manual Grader <br> *Architecture Grading*] --> F
        H --> I[Candidate Certification Service <br> *Generate Certification*]
        B --> |Viewer| J[Certification Viewer Service <br> *View Certification*]

        F --> L[(Candidate Status DB <br> *see Test 1*)]
        G --> M[(Submission Ungraded DB)]
        H --> N[(Grade and Feedback DB)]
        I --> O[(Certification DB)]

        H --> P[Candidate Architecture Notification <br> *Candidate Architecture Results*]
        P --> Q[Email]
        P --> R[Certificate]

        O --> J
```

**Problem:**

- Manual grading of architecture submissions takes approximately 8 hours per candidate.
- The expected overall increase in volume (5–10×) in a short term expansion and the 21% increase over the next 4 years will strain the expert resources.
- The architecture submissions are complex and nuanced, making fully automated grading risky.

> - In addition to grading, it would be useful some help in providing detailed, actionable feedback about the candidate’s submission. Producing such feedback manually is time-consuming and maybe prone to inconsistencies.

Reviewer 2: ```suggestion
- In addition to grading, it would be useful to offer some help in providing detailed, actionable feedback about the candidate’s submission. Producing such feedback manually is time-consuming and maybe prone to inconsistencies.
```

- Grading accuracy and transparency is critical because mistakes can result in serious damage to the candidates' careers and to Certifiable, Inc.'s reputation.

---

## 3. Decision

**Chosen Approach:**

Implement an AI-Assisted Grading Service for architecture submissions that:

- Uses a private deployment of an LLM model or an LLM API from an external provider - both multimodal with text and vision capabilities.
- **Grades the candidate's submission:** With the help of AI generates a preliminary grade based on the candidate's submission.

> - **Generates Integrated Feedback:** Produces detailed, AI-generated feedback that highlights the positives and the negatives in the candidate’s submission. This pre-filled feedback covers areas such as missing components, deviations from best practices, and structural issues.

Reviewer 3: ```suggestion
- **Generates Integrated Feedback:** Produces detailed, AI-generated feedback that highlights the positives and the negatives in the candidate’s submission. This pre-filled feedback analyses the submission using the grading guidelines as a basis and covers areas such as missing components, deviations from best practices, and structural issues.
```

    J -->M[Email Results] --> A
```

**Changed Certification Testing Architecture - Test 2: Architecture Solution:**

Changed components:

- The Architecture Grader will be changed to use the LLM Gateway to generate the grades, feedback and confidence scores. It will be changed to support the changes to the Expert Grading User Interface. Additionally, it will be changed to use the Audit Service to support the new auditing features.
- The Expert Grading User Interface will be updated to allow the experts to review and refine the new AI-generated grades and feedback while also presenting the confidence scores.

New components:

- The LLM Gateway will provide the LLM services to other components. The component relevant in this case is the Architecture Grader. This service will allow to switch between model providers, in-house or external.
- The LLM Model Serving represents in-house or external LLM models. These can be fine-tuned and/or prompt according to the specific needs of the architecture submissions. Its actual design and implementation depends on the Research and Prototype of the [Implementation Plan](#8-implementation-plan) phase. It will be used by the LLM Gateway.

> - The Audit Service will provide the necessary auditing features to support the afore-mentioned improvement activities.

Reviewer 3: We should be consistent in both our RFCs.... This persistence could/should be the same in both systems.
```suggestion
- The Data store will provide the necessary auditing features to support the afore-mentioned improvement activities.
```

   - Together with a team of selected experts, the project team will conduct manual tests using available models and existing golden examples to determine the best approach to grading and feedback generation. This step is essential to validate the proposed solution and ensure that it meets the required quality standards.

   - Given the latest developments we believe that fine-tuning a pre-trained like Llama 3.2 Vision or using an LLM provider like Google Gemini or other that can deliver identical capabilities would be good approaches. As stated later, we're assuming there are guidelines and golden examples that can be used.

   - We don't eliminate the possibility of using the existing models (opensource or proprietary) without fine-tuning and relying only on prompting techniques. The decision to fine-tune or not should become clear during this phase.

   - The confidential requirements should be taken into account when choosing the model, the provider, in-house vs external and so on. The team should also consider the costs and the infrastructure needed to support the chosen model.

>    - One critical outcome from this phase is also a set of test cases as exhaustive as necessary that can be used to validate the system, both from the start but also during future changes. Those tests should also be reevaluated in the future when modifying existing case studies or adding new ones.

Reviewer 3: Tests also to enable benchmarking different models

Reviewer: ```suggestion
   - One critical outcome from this phase is also a set of test cases as exhaustive as necessary that can be used to validate the system, both from the start and during future changes. Those tests should also be reevaluated in the future when modifying existing case studies or adding new ones. Tests should also enable benchmarking different models.
```

   - Develop a prototype that conforms to generate both preliminary grades and detailed draft feedback insights using historical graded submissions as benchmarks.

2. **Expert Grading User Interface Re-design:**

   The existing Expert Grading User Interface will be re-designed to support the new AI-assisted grading and feedback generation features.

>    - The candidate’s architectural submission side-by-side to the “golden standard.”

Reviewer 3: in the Readme we referenced guidelines, should we reference the golden standard as well?

Reviewer: Done
